Error behaviors (including `onError: "NULL"`) (#1163)
graphql/graphql-spec: spec/Section 4 -- Introspection.md

## Schema Introspection

The schema introspection system is accessible from the meta-fields `__schema`,
`__type` and `__service` which are accessible from the type of the root of a
query operation.

```graphql
__schema: __Schema!
__type(name: String!): __Type

> __service: __Service!

Author: Discussed at WG, conclusion was:

- we do not need to include this in the spec
- we should never add anything that requires runtime introspection of service capabilities to the spec - build time should always suffice
- **if someone is disabling introspection in production, there's a very good chance that they're doing the wrong thing and should be using Trusted Documents instead**

  NON_NULL
}

enum __ErrorBehavior {
  NO_PROPAGATE
  PROPAGATE

>   ABORT

Reviewer: Moving my [previous comment](https://github.com/graphql/graphql-spec/pull/1153#discussion_r2016918168) here, love this proposal overall but I would love it even more with `NULL`

```graphql
enum __ErrorBehavior {
  NULL
  PROPAGATE
  ABORT
}
```

Author: To summarize my thoughts on it from the previous thread: `NULL` is what I originally wrote, I agree it looks and initially feels nicer; but ultimately I discounted it for a number of reasons:

### Communication issues

If you tell someone on the phone to set `onError: NULL`, they're very likely to instead set `onError: null`.

### Ambiguity issues

`onError: null` and `onError: NULL` are very similar, but the first means `PROPAGATE` and the latter means `NO_PROPAGATE`. This is likely to be the source of many user errors and much frustration (both for newbies, and for the people supporting them including library authors).

### Errors _already_ become null in the spec:

> [an error] raised [...] is handled as though the response position [...] resolved to `null`
> -- https://spec.graphql.org/draft/#sel-FANVNJCAACGW-qR

`onError: NULL` doesn't implicitly seem to mean something different from this.

### Please suggest other alternatives!

I'm very open to changing the word `NO_PROPAGATE` to something else, I really don't like `NO_PROPAGATE`, not least because I always want to spell it propOgate rather than propAgate. But I think changing it to `NULL` would be a mistake.

I've gone through about 30 alternatives when brainstorming it, and none are sufficiently clear whilst also being shorter than `NO_PROPAGATE`. My favourite alternatives are things like `LOCAL`/`LOCALIZE`/`ISOLATE`/`CONTAIN`, but I'm not sure they convey the meaning quite as well as `NO_PROPAGATE` does.

Author: Ooo... Maybe `LOCAL_NULL` actually is quite compelling. Not thought of that one before.

  INPUT_OBJECT
  INPUT_FIELD_DEFINITION
}

type __Service {

>   capabilities: [__Capability!]!

Author: I've thought about this quite a bit, and I think capabilities is better - it's more indicative of intent/purpose.

The aim is that clients can use the "capabilities" of the service to inform their own behaviors in an automated way, such that end users can just point to a GraphQL endpoint and the client will take care of the rest - they don't need to configure it extensively (and the configuration can be automatically updated on each release without frontend engineers needing to do anything).

For example:

- does it have the capability of handling fragment arguments? If not, I must transpile them.
- does it have the capability of handling document directives? If not, I must find them invalid
- does it have the capability of customizing error propagation behavior? If not, I must assume that it will propagate errors (and should not pass `onError`)
- does it support subscriptions over websockets? SSE? If so, what are the endpoints?
- does it support automated document persistence? If so, which protocol does it use and what are the relevant configuration parameters?
- does it support notification of new schema releases? If so, at which endpoint can I subscribe to these events? (Especially useful for GraphiQL)
- does it support incremental delivery? Which version?
- does it support {RFC feature X}?
- etc

The capabilities system _can_ be used for metadata, but that's not its purpose; and calling it metadata significantly changes what is being proposed (and also encourages it to be used for feeding a huge amount more "metadata" into the system). Capabilities inform automated client decisions and actions; metadata is ancillary and often intended for human consumption.

I will try and ensure this is correctly reflected in the spec text.

    {true}, deprecated arguments are also returned.
- `isRepeatable` must return a Boolean that indicates if the directive may be
  used repeatedly at a single location.

### The \_\_Service Type

The `__Service` type is returned from the `__service` meta-field and provides
information about the GraphQL service, most notably about its capabilities. This
type was added after the original release of GraphQL, so older schemas may not

> support it.

Reviewer: Do we really need the `__service` indirection vs having `__capabilities`? 
My mental model is that the endpoint is the service. It has a `__schema` and `__capabilities` are at the same level (except they can't be expressed in terms of type system).

Reviewer: Yikes, I'm having an existential crisis now...

Can clients know whether `__service` is supported? As a meta field, it's not returned in introspection, right? So it's too late now, we have to retrofit everything inside `__schema`???

Fields\:

- `capabilities` must return a list of `__Capability` indicating the
  capabilities supported by the service.

### The \_\_Capability Type

:: A _capability_ describes a feature supported by the GraphQL service but not
directly expressible in the type system. This may include experimental GraphQL
features (such as new syntax or behavior), protocol support (such as GraphQL
over WebSockets), or additional operational metadata (such as release
identifiers). Capabilities may be supplied by the GraphQL implementation, the
service, or both.

The `__Service.capabilities` field exposes a _capability_ list. A _capability_
consists of a capability identifier and optionally a string value.

**Capability identifier**

> A capability identifier is a string value composed of two or more segments

Reviewer 2: Clarify that this is a suggestion

Author: Governed by the "preserve option value" guiding principle:

> It's hard to know what the future brings; whenever possible, decisions should be made that allow for more options in the future. Sometimes this is unintuitive: spec rules often begin more strict than necessary with a future option to loosen when motivated by a real use case.

I think we should keep this as a requirement rather than a suggestion/recommendation. It's quite simple to comply with, and we can always loosen it up in future if need be.

E.g. maybe we might want to add an SDL syntax for this in future, it would be a shame to have to wrap all the keys in strings:

```
service {
  capabilities {
    org.graphql.onError
    org.graphql.defailtErrorBehavior: "PROPAGATE"
    org.graphql.operationDescription
    org.graphql.ws.subscriptions.endpoint: "wss://example.com/graphql"
    org.graphql.ws.queries
    org.graphql.ws.mutations
    com.graphile.env: "Production"
    com.graphile.hash: "sdf9009s8dg09sdf809"
    org.graphql.rfc.ccn: "v1"
    org.graphql.rfc.incrementalDelivery: "June2023"
  }
}
```

That said, I'm not too bothered and happy to remove it if there's a good reason.

separated by a period (`.`). Each segment must begin with an ASCII letter, and
must contain only ASCII letters, digits and hyphens. These constraints are
inspired by reverse domain name notation to encourage global uniqueness and

> collision-resistance. Unlike the domain name system, capability identifiers are

Reviewer: I like this!

begin with a prefix derived from a DNS name they control (e.g.,
{"com.example."}).

Clients should compare identifiers using exact string equality and should ignore
unknown identifiers.

Implementers should not change the meaning of capability identifiers; instead, a
new capability identifier should be used when the meaning changes. Implementers
should ensure that capability identifiers remain stable and version-agnostic
where possible; capability versioning, if needed, can be indicated using dot
suffixes (e.g.{ "org.example.capability.v2"}).

This system enables incremental feature adoption and richer tooling
interoperability, while avoiding tight coupling to specific implementations.

**Capability value**

The value assigned to a given capability may either be {null} to simply indicate
the capability is supported, or a string to provide additional information. For
example {"org.graphql.onError"} does not require additional information and thus
uses the value {null}, whereas {"org.graphql.defaultErrorBehavior"} needs the
{value} to indicate which _error behavior_ is the default.

**Specified capabilities**

This version of the specification defines the following capabilities:

> - {"org.graphql.defaultErrorBehavior"} - indicates the _default error behavior_
>   of the service via the {value}. If not present, must be assumed to be
>   {"PROPAGATE"}.
> - {"org.graphql.onError"} - indicates that the service allows the client to
>   specify {onError} in a request to indicate the _error behavior_ the service
>   should use for the request. {value} is {null}.

Reviewer: What about this?

```suggestion
- {"org.graphql.defaultErrorBehavior"} - indicates the _default error behavior_
  of the service via the {value}. 
  If present, the service must be assumed to support the {onError} parameter 
  to indicate the _error behavior_ on requests.
  If not present, must be assumed to be {"PROPAGATE"}.
```

Less invalid cases possible.

Author: This requires services to support the `onError` request parameter forever, I don't think we should do that. If a service wants to start up that only supports the `NO_PROPAGATE` error behavior and doesn't want to allow clients to influence that, I think it should be allowed to.

Reviewer: SGTM 👍
